src/monodroid/jni/basic-utilities.hh

```diff
@@ -186,69 +178,34 @@ namespace xamarin::android
 #endif // def WINDOWS
 		bool            is_path_rooted (const char *path);
 
-		template<typename Ret, typename P1, typename P2>
-		inline Ret add_with_overflow_check (const char *file, uint32_t line, P1 a, P2 b) const
+		template<typename StringType = char>
+		size_t calculate_length (const StringType* s)
 		{
-			Ret ret;
-
-			if (XA_UNLIKELY (__builtin_add_overflow (a, b, &ret))) {
-				log_fatal (LOG_DEFAULT, "Integer overflow on addition at %s:%u", file, line);
-				exit (FATAL_EXIT_OUT_OF_MEMORY);
-				return static_cast<Ret>(0);
-			}
-
-			return ret;
+			return strlen (s);
 		}
 
-		// Can't use templates as above with add_with_oveflow because of a bug in the clang compiler
-		// shipped with the NDK:
-		//
-		//   https://github.com/android-ndk/ndk/issues/294
-		//   https://github.com/android-ndk/ndk/issues/295
-		//   https://bugs.llvm.org/show_bug.cgi?id=16404
-		//
-		// Using templated parameter types for `a` and `b` would make clang generate that tries to
-		// use 128-bit integers and thus output code that calls `__muloti4` and so linking would
-		// fail
-		//
-		template<typename Ret>
-		inline Ret multiply_with_overflow_check (const char *file, uint32_t line, size_t a, size_t b) const
+		template<typename StringType = char, typename ...Strings>
+		size_t calculate_length (const StringType* s1, Strings... strings)
 		{
-			Ret ret;
+			assert_char_type<StringType> ();
 
-			if (XA_UNLIKELY (__builtin_mul_overflow (a, b, &ret))) {
-				log_fatal (LOG_DEFAULT, "Integer overflow on multiplication at %s:%u", file, line);
-				exit (FATAL_EXIT_OUT_OF_MEMORY);
-				return static_cast<Ret>(0);
-			}
-
-			return ret;
+			return strlen (s1) + calculate_length (strings...);
```

Comment:
Should this also have a compiler assert to prevent overflow here?

Reply:
you mean `ADD_WITH_OVERFLOW_CHECK`? If yes, then I'd rather avoid it as this code could generate a number of calls to the built-in thus killing the performance of this function. The likelihood that it will ever overlfow is very, very, very low considering we don't merge user-provided strings with this function. It's used to merge path names with literal strings, mostly. Let's not overengineer... :)